### PR TITLE
Chore: explicitly list out all Expr types in TypeCoercionRewriter::mutate

### LINKED ISSUE #8990: explicitly list out all Expr types in `TypeCoercionRewriter::mutate`
It might make sense to remove the catch all  from this `match` expression

https://github.com/apache/arrow-datafusion/blob/bc0ba6a724aaaf312b770451718cbce696de6640/datafusion/optimizer/src/analyzer/type_coercion.rs#L413

And explicitly list out all Expr types so that we don't accidentally another Expr type in the future

Is that something you might be willing to do in another PR @guojidan  ?

_Originally posted by @alamb in https://github.com/apache/arrow-datafusion/pull/8982#discussion_r1464993523_

### PR DESCRIPTION
**Pull request by guojidan**
## Which issue does this PR close?

Closes #8990 .

### CODE REVIEW

**Comment by guojidan** [comment]
indexmap::IndexSet crate will deprecate `remove` method at [2.2.0](https://docs.rs/indexmap/2.2.0/indexmap/set/struct.IndexSet.html#method.remove) version, but we use [2.0.0](https://github.com/apache/arrow-datafusion/blob/main/Cargo.toml#L63), why test cases fail? 😕 

and `remove` methed just a wrap of `swap_remove` method, Can we directly use `swap_remove`? 🤔

**Comment by alamb** [comment]
The CI failure was fixed in https://github.com/apache/arrow-datafusion/pull/9034

**Review by alamb — APPROVED** [review]
Thank you @guojidan  -- this is a nice improvement

File: datafusion/optimizer/src/analyzer/type_coercion.rs [hunk 1]
```diff
@@ -414,7 +414,22 @@ impl TreeNodeRewriter for TypeCoercionRewriter {
...
             }
-            expr => Ok(expr),
+            Expr::Alias(_)
+            | Expr::Column(_)
+            | Expr::ScalarVariable(_, _)
+            | Expr::Literal(_)
+            | Expr::SimilarTo(_)
+            | Expr::IsNotNull(_)
+            | Expr::IsNull(_)
+            | Expr::Negative(_)
```
Comment (alamb) [inline datafusion/optimizer/src/analyzer/type_coercion.rs:424]: 🤔  looks like there are some other bugs here;

```
❯ select -'100';
Internal error: Can't create negative physical expr for (- 'Literal { value: Utf8("100") }'), the type of child expr is Utf8, not signed numeric.
This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
❯ select -'100'::int;
+-----------------+
| (- Utf8("100")) |
+-----------------+
| -100            |
+-----------------+
```

I will file another ticket

**Review by alamb — APPROVED** [review]
Thanks @guojidan  -- this is awesome. Even better it helped find another bug 🎉

**Pull request merged**